### PR TITLE
Implement Gtk CSS utility with button implementation

### PR DESCRIPTION
**Pull request by ArchKudo**
## May 2019 update
- [x] Add Gtk CSS utility (w/ documentation and tests)
- [x] Added Button implementation
  - [x] Removed `set_enabled` from toga_gtk `Button` class as it is already implemented in `Widget` (See PR #703)

## References for reviewers:
https://developer.gnome.org/gtk3/stable/chap-css-overview.html
https://developer.gnome.org/gtk3/stable/chap-css-properties.html

## About
Most styling functions for Gtk have been depreciated since Gtk+3.16([Example](https://lazka.github.io/pgi-docs/#Gtk-3.0/classes/Widget.html#Gtk.Widget.override_background_color)), and ~~are~~ can be now modified using Gtk CSS
This PR is thus an attempt to add styling to Gtk widgets using CSS 

## What this should eventually include / Updates:
- [x] CSS helper methods
- [ ] ~Use CSS instead of Pango for changing fonts~
- [x] Document changes
- [x] Widgets implementation like `set_font`, `set_color`, ...
- [x] Extensively unittest old and new code
- [x] I also took the liberty to format files and remove unnecessary imports :innocent: :open_hands: 

## Current Roadblocks
*  ~A certain test to set Gtk property is failing on ci, my guess is due to different version of pygobject being installed on ci, Could try again after PR #374 gets merged~
* ~Can only change `font_family` and `font_size` through interface: See [gist](https://gist.github.com/ArchKudo/7265517c4e0e23dfe37cad098145a01a) for details~
* ~According to what I understood from chatting on Gtk+ irc, `background_color` cannot be unittested, since it is no more a simple `GDK.RGBA`  type but can be anything including image, video, cairo object, etc.~

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: src/gtk/toga_gtk/libs/utils.py [hunk 1]
```diff
@@ -15,3 +17,41 @@ def gtk_alignment(alignment):
...
+    Generate Gtk CSS from property-value pairs for a single selector
+    Args:
+    selector(str): Properties are applied to this CSS selector
+    decl(dict): CSS property-value pairs
+    Returns:
+    str: Gtk CSS rule
+    """
+    res = selector + ' {\n'
+    for attr, val in decls.items():
+        res += f'\t{attr}: {val};\n'
```
Comment (ArchKudo) [inline src/gtk/toga_gtk/libs/utils.py]: Just realized f-strings are Python3.6+ feature, I should have use `.format()`

File: src/gtk/tests/libs/test_utils.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import unittest
+gi.require_version('Gtk', '3.0')
+from gi.repository import Gtk, Gdk
+from toga_gtk.libs.utils import css_rule_factory, gtk_apply_css
+
+
+class TestCSSRuleFactoryMethod(unittest.TestCase):
+    def setUp(self):
+        self.single_line_property = {'color': '#f00'}  # Red
+        self.multi_line_property = {
```
Comment (ArchKudo) [inline src/gtk/tests/libs/test_utils.py]: Just realized dicts don't have order, causing tests to fail :cry:

File: src/gtk/tests/libs/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
+import sys
+from collections import OrderedDict
+import unittest
+try:
+    import gi
+    gi.require_version('Gtk', '3.0')
+    from gi.repository import Gtk, Gdk
+except ImportError:
+    import sys
+    # If we're on Linux, Gtk *should* be available. If it isn't, make
```
Comment (brutusthebee) [inline src/gtk/tests/libs/test_utils.py]: At column 5: (F811) redefinition of unused 'sys' from line 1

File: src/gtk/tests/libs/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        self.single_line_property = {'color': '#f00'}  # Red
+        self.multi_line_property = {
+            'font-family': 'sans',
+            'font-size': '12px',
+            'font-style': 'italic',
+        }
+
+        # Use OrderedDict for Python < 3.6
+        if sys.version_info < (3,6):
+            d = OrderedDict()
```
Comment (brutusthebee) [inline src/gtk/tests/libs/test_utils.py]: At column 33: (E231) missing whitespace after ','

File: src/gtk/tests/libs/test_utils.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    def test_empty_selector_raises_value_error(self):
+        with self.assertRaises(ValueError):
+            css_rule_factory('', self.single_line_property)
+
+    def test_empty_property_raises_value_error(self):
+        with self.assertRaises(ValueError):
+            css_rule_factory('button', {})
+
+@unittest.skipIf(Gtk is None, "Can't run GTK implementation tests on a non-Linux platform")
+class TestGTKApplyCSSMethod(unittest.TestCase):
```
Comment (brutusthebee) [inline src/gtk/tests/libs/test_utils.py]: At column 1: (E302) expected 2 blank lines, found 1

File: src/gtk/toga_gtk/libs/utils.py [hunk 6]
```diff
@@ -1,5 +1,7 @@
-from toga.constants import *
-
+import gi
+gi.require_version('Gtk', '3.0')
+from gi.repository import Gtk
```
Comment (brutusthebee) [inline src/gtk/toga_gtk/libs/utils.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/libs/utils.py [hunk 7]
```diff
@@ -1,5 +1,7 @@
...
+from toga.constants import LEFT, RIGHT, CENTER, JUSTIFY
```
Comment (brutusthebee) [inline src/gtk/toga_gtk/libs/utils.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/libs/utils.py (hunk 6 above)
Comment (brutusthebee) [inline src/gtk/toga_gtk/libs/utils.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/libs/utils.py (hunk 7 above)
Comment (brutusthebee) [inline src/gtk/toga_gtk/libs/utils.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/widgets/button.py [hunk 8]
```diff
@@ -1,26 +1,23 @@
...
         self.native.set_label(self.interface.label)
         self.rehint()
 
-    def set_enabled(self, value):
-        self.native.set_sensitive(value)
-
     def set_background_color(self, value):
-        self.interface.factory.not_implemented('Button.set_background_color()')
+        if value:
+            gtk_apply_css(self.native, {"background-color": value})
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/button.py]: Is there a reason that this is an external utility method, rather than a method on the base toga-GTK widget?

**Comment by ArchKudo** [comment]
> In particular, I'm not clear (from the documentation you linked, combined with the usage here) what the lifecycle of a GTK style provider/context is intended to be. The CSS language GTK has implemented includes widget and class specifiers, but the API (especially as used here) seems to be specifically targeted at individual widgets.
> So - what's the intended lifecycle here? Are you meant to apply a stylesheet to an entire application? Or is per-widget the intended usage? 
>
The recommended way is to have a single application level CSS as we would have while creating a website, so in that sense toga doesn't exactly need it, since, it implements its own styling.. however, styling using python api was limited with GTK3 as, some of the theming API was deprecated in favour of CSS..

>And if so - is the style applied to the "button" target on each button, or on "button#id-1234"? (i.e., what is the interaction of CSS's element targetting namespace and GTK's per-widget style application API?) 

It applies to the specific widget only but, it is totally possible to have it the other way around depending on the selector...
 
> Is the only way to specify styles to load-from-string? Do we really have to pay a text parsing overhead every time we want to change a style attribute on a widget? That seems... wasteful.

Given the underlying API is in C, I can't find any other method either than to load CSS from a file or as a string..  :woman_shrugging:
 
> Are styles additive? Or does a new style declaration replace an old one?
> 
> If they're additive (as would be suggested by the `add_provider()` API naming - can you modify an existing style context?

Yes they are, and follow the CSS precedence rules, also setting priority to application always applies the most current change..
 
> I'm not denying that this code works. I just want to make sure we're actually using the GTK API the right way - because the way it's being used right now, it seems like there's going to be a lot of object creation going on as you construct CSSProviders and get StyleContexts. And, if you were to switch (for example) the color of a widget back and forth between two values, the internal style definition of the GTK widget is going to become a long list of internal overrides.

Oh! Totally agree, it probably didn't cross my mind, one solution I can think of is adding a class based UID to each CSS rule since a quick search on [SO](https://stackoverflow.com/questions/47607420/gtk3-how-to-reset-the-background-color-of-a-widget-to-default-state-after-using) says there exists API to remove a css class though, I must admit this requires some more research on my side best not to rush it.. 

Do let me know if any of my responses aren't clear enough and, thanks a lot for reviewing!

**Comment by freakboy3742** [comment]
> The recommended way is to have a single application level CSS as we would have while creating a website, so in that sense toga doesn't exactly need it, since, it implements its own styling.. however, styling using python api was limited with GTK3 as, some of the theming API was deprecated in favour of CSS..

In which case - are there any performance considerations here? Is setting lots of individual per-widget styles any faster or slower than having a single application-level style?
 
> > And if so - is the style applied to the "button" target on each button, or on "button#id-1234"? (i.e., what is the interaction of CSS's element targetting namespace and GTK's per-widget style application API?)
> 
> It applies to the specific widget only but, it is totally possible to have it the other way around depending on the selector...

Again - what are the performance considerations? Is it faster to set:
* `{ background-color: red }` set as a property on a specific widget (i.e.,  a global style definition, but widget-scoped by virtue of being invoked on a specific widget)
* `button: { background-color: red }` set as a property on a specific widget (i.e.,  a global widget style definition, but widget-scoped by virtue of being invoked on a specific widget)
* `#id-12345: { background-color: red }` set as a property on a specific widget (strictly overspecified due to being invoked on a specific widget)
* `button#id-12345: { background-color: red }` set as a property on a specific widget (doubly-overspecified due to being invoked on a specific widget)
* `#id-12345: { background-color: red }` set as a global style
* `button#id-12345: { background-color: red }` set as a global style (overspecified because there is only one widget in an app with a given ID)

I'd be interested in seeing a performance comparison using a test-bed application - for example, an app with hundreds of buttons in a window, repeatedly changing a style back and forth on *one* of the buttons. This doesn't need to be a Toga application (in fact, it's probably better if it isn't - we're looking to verify the specific performance characteristics of GTK, not the Toga overhead).

> > Are styles additive? Or does a new style declaration replace an old one?
> > If they're additive (as would be suggested by the `add_provider()` API naming - can you modify an existing style context?
> 
> Yes they are, and follow the CSS precedence rules, also setting priority to application always applies the most current change..

I'm intrigued by the performance characteristics here, too. Is the internal structure retaining the full style definition and recomputing the "net" style after every addition? Or is it only retaining the "current" state of the style, and when you  load a new style, determining if selectors affect the "current" style. Or something in between?

The specifics of the implementation don't particularly matter - but if it's the former, I'd expect to see a performance difference between iteration 0-100 and iteration 100000 and 100100.

> Oh! Totally agree, it probably didn't cross my mind, one solution I can think of is adding a class based UID to each CSS rule since a quick search on [SO](https://stackoverflow.com/questions/47607420/gtk3-how-to-reset-the-background-color-of-a-widget-to-default-state-after-using) says there exists API to remove a css class though, I must admit this requires some more research on my side best not to rush it..

I can see how using a "unique class" as an identifier could be manipulated to our advantage - however, the use case isn't likely to be "reset style to default value", it's "overwrite existing style with new value". To that end, it's probably worth adding `button.id-12345: { background-color: red }` to the pool of options to check in the performance test.

**Comment by phildini** [comment]
Hi there! It looks like this PR might be dead, so we're closing it for now. Feel free to re-open it if you'd like to continue. 😄

**Comment by danyeaw** [comment]
@ArchKudo This PR was a really good step towards us getting support for Button and Label font sizes, and I think you were like 90% of the way there. Is this something you are still interested in finishing? We have a sprint going on this weekend. Otherwise we could have someone create a new PR from where you left off.

**Pull request closed without merging**